test/common/upstream/wrsq_scheduler_test.cc

```diff
@@ -0,0 +1,289 @@
+    auto p1 = sched1.pickAndAdd({});
+    auto p2 = sched2.pickAndAdd({});
+    EXPECT_EQ(picks[rounds], *p1);
+    EXPECT_EQ(*p2, *p1);
+  }
+}
+
+// Expire all objects and verify nullptr is returned.
+TEST(WRSQSchedulerTest, ExpireAll) {
+  Random::MockRandomGenerator random;
+  WRSQScheduler<uint32_t> sched(random);
+
+  // The weights are small enough that we can just burn through all the relevant random numbers that
+  // would be generated as long as we hit 12 consecutive numbers for each part.
+  uint32_t rnum{0};
+
+  {
+    // Add objects of the same weight.
+    auto e1 = std::make_shared<uint32_t>(42);
+    auto e2 = std::make_shared<uint32_t>(37);
+    sched.add(1, e1);
+    sched.add(1, e2);
+
+    {
+      auto e3 = std::make_shared<uint32_t>(7);
+      auto e4 = std::make_shared<uint32_t>(13);
+      sched.add(5, e3);
+      sched.add(5, e4);
+
+      // We've got unexpired values, so we should be able to pick them. While we're at it, we can
+      // check we're getting objects from both weight queues.
+      uint32_t weight1pick{0}, weight5pick{0};
+      for (int i = 0; i < 1000; ++i) {
+        EXPECT_CALL(random, random()).WillOnce(Return(rnum++));
+        switch (*sched.pickAndAdd({})) {
+        case 42:
+        case 37:
+          ++weight1pick;
+          break;
+        case 7:
+        case 13:
+          ++weight5pick;
+          break;
+        default:
+          EXPECT_TRUE(false) << "bogus value returned";
+        }
+      }
+      EXPECT_GT(weight5pick, 0);
+      EXPECT_GT(weight1pick, 0);
+    }
+
+    // Expired the entirety of the high-probability queue. Let's make sure we behave properly by
+    // expiring them and only returning the unexpired entries.
+    for (int i = 0; i < 1000; ++i) {
+      EXPECT_CALL(random, random()).WillRepeatedly(Return(rnum++));
+      switch (*sched.peekAgain({})) {
+      case 42:
+      case 37:
+        break;
+      default:
+        EXPECT_TRUE(false) << "bogus value returned";
+      }
+    }
+  }
+
+  // All values have expired, so only nullptr should be returned.
+  EXPECT_CALL(random, random()).WillRepeatedly(Return(rnum++));
+  EXPECT_EQ(sched.peekAgain({}), nullptr);
+}
+
+// Validate that a new requested weight is honored.
+TEST(WRSQSchedulerTest, ChangingWeight) {
+  NiceMock<Random::MockRandomGenerator> random;
+  WRSQScheduler<uint32_t> sched(random);
+
+  auto e1 = std::make_shared<uint32_t>(123);
+  auto e2 = std::make_shared<uint32_t>(456);
+  auto e3 = std::make_shared<uint32_t>(789);
+  sched.add(1, e1);
+  sched.add(0, e2);
+
+  std::function<double(const uint32_t v)> f = [e1, e2](const uint32_t v) {
+    if (v == *e1) {
+      return 0.0;
+    } else {
+      return 1.0;
+    }
+  };
+
+  // Expecting only e1 to be picked.
+  for (uint32_t rounds = 0; rounds < 128; ++rounds) {
+    auto peek = sched.peekAgain({});
+    auto p = sched.pickAndAdd({});
+    EXPECT_EQ(*e1, *p);
+    EXPECT_EQ(*peek, *p);
+  }
+
+  auto p = sched.pickAndAdd(f);
```

Comment:
Please add a comment here describing what the weights look like after this, maybe something like "After this, weights are e1 -> 0, e2 -> 0, e3 -> 1"